### PR TITLE
modify segment generator to not return illegal filenames

### LINKED ISSUE #6772: [Bug] Segment name doesn't support  '/'
Since we support segment name with start/end time in it, if the time format is in `yy/mm/dd` then it introduces special characters like '/' into segment name.

Segment upload path if fine, we can see the segment is loaded into the controller, but it's stored into a directory format.
E.g. segment name: `covid_cases_OFFLINE_2020/10/23_2021/01/28_0` is stored into many levels of directory:
```
➜ find pinot-docker-demo/pinot/controller/data/Deleted_Segments/covid_cases
pinot-docker-demo/pinot/controller/data/Deleted_Segments/covid_cases
pinot-docker-demo/pinot/controller/data/Deleted_Segments/covid_cases/covid_cases_OFFLINE_2020
pinot-docker-demo/pinot/controller/data/Deleted_Segments/covid_cases/covid_cases_OFFLINE_2020/10
pinot-docker-demo/pinot/controller/data/Deleted_Segments/covid_cases/covid_cases_OFFLINE_2020/10/23_2021
pinot-docker-demo/pinot/controller/data/Deleted_Segments/covid_cases/covid_cases_OFFLINE_2020/10/23_2021/01
pinot-docker-demo/pinot/controller/data/Deleted_Segments/covid_cases/covid_cases_OFFLINE_2020/10/23_2021/01/28_0
```

From zk side, the segment is also stored into many levels:
image

Then this prevents segment download from pinot server(I don't even bother this part as it's already pretty wrong from controller and zk side.)

The log is:
```
pinot-server_1      | Got temporary error status code: 500 while downloading segment from: http://pinot-controller:9000/segments/covid_cases/covid_cases_OFFLINE_2020%2F10%2F23_2021%2F01%2F28_0 to: /tmp/data/server/data/segment/covid_cases_OFFLINE/tmp-covid_cases_OFFLINE_2020/10/23_2021/01/28_0-b65e3a35-d867-4a83-8404-707172c18e4e/covid_cases_OFFLINE_2020/10/23_2021/01/28_0.enc
pinot-server_1      | org.apache.pinot.common.exception.HttpErrorStatusException: Got error status code: 500 (Internal Server Error) with reason: "Failed to get reason" while sending request: http://pinot-controller:9000/segments/covid_cases/covid_cases_OFFLINE_2020%2F10%2F23_2021%2F01%2F28_0
pinot-server_1      | 	at org.apache.pinot.common.utils.FileUploadDownloadClient.downloadFile(FileUploadDownloadClient.java:670) ~[pinot-all-0.7.0-SNAPSHOT-jar-with-dependencies.jar:0.7.0-SNAPSHOT-04e2bcce5026f5911183a27f5e8fed259572dfaf]
```

### PR DESCRIPTION
**Pull request by walterddr**
## Description
Add logic in `SegmentNameGenerator`s to no allow illegal filename characters
This fixes #6772.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
Nope

Does this PR fix a zero-downtime upgrade introduced earlier?
Nope

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options - No
- Deprecation of configurations - No 
- Signature changes to public methods/interfaces - No
- New plugins added or old plugins removed - No 

## Documentation
N/A

### CODE REVIEW

**Review by Jackie-Jiang — COMMENTED** [review]
Would suggest fixing it by rejecting the invalid table name / prefix / postfix instead of replacing the invalid characters just in the segment name. Having invalid character within table name can cause other issue even after fixing the segment name

**Comment by walterddr** [comment]
> Would suggest fixing it by rejecting the invalid table name / prefix / postfix instead of replacing the invalid characters just in the segment name. Having invalid character within table name can cause other issue even after fixing the segment name

thanks for the review and I like this idea. just to clarify do we want to reject only or do you think we should also apply the auto-replace strategy to table name / prefix / postfix as well, like what I did in this PR for segment name?

**Comment by Jackie-Jiang** [comment]
@walterddr I think we should reject and throw exception instead of trying to handle it. We can validate the table name and date time format when creating the table

**Comment by walterddr** [comment]
yup I thought the same so already implemented. plz kindly take a look

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/NormalizedDateSegmentNameGenerator.java [hunk 1]
```diff
@@ -52,6 +52,10 @@
   public NormalizedDateSegmentNameGenerator(String tableName, @Nullable String segmentNamePrefix,
       boolean excludeSequenceId, @Nullable String pushType, @Nullable String pushFrequency,
       @Nullable DateTimeFormatSpec dateTimeFormatSpec) {
+    Preconditions.checkArgument(
+        tableName != null && !tableName.matches(INVALID_SEGMENT_NAME_REGEX));
+    Preconditions.checkArgument(
+        segmentNamePrefix == null || !segmentNamePrefix.matches(INVALID_SEGMENT_NAME_REGEX));
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/NormalizedDateSegmentNameGenerator.java:58]: Let's check the `_segmentNamePrefix` instead

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/SimpleSegmentNameGenerator.java [hunk 2]
```diff
@@ -39,12 +40,20 @@
...
+    Preconditions.checkArgument(
+        segmentNamePostfix == null || !segmentNamePostfix.matches(INVALID_SEGMENT_NAME_REGEX));
     _segmentNamePrefix = segmentNamePrefix;
     _segmentNamePostfix = segmentNamePostfix;
   }
 
   @Override
   public String generateSegmentName(int sequenceId, @Nullable Object minTimeValue, @Nullable Object maxTimeValue) {
+    Preconditions.checkArgument(
+        minTimeValue == null  || !minTimeValue.toString().matches(INVALID_SEGMENT_NAME_REGEX));
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/SimpleSegmentNameGenerator.java:54]: (code style) extra space?

File: pinot-segment-spi/src/test/java/org/apache/pinot/segment/spi/creator/name/NormalizedDateSegmentNameGeneratorTest.java [hunk 3]
```diff
@@ -20,20 +20,24 @@
...
+  private static final String MALFORMED_TABLE_NAME = "my/Table";
   private static final String SEGMENT_NAME_PREFIX = "myTable_daily";
+  private static final String MALFORMED_SEGMENT_NAME_PREFIX = "myTable\\daily";
   private static final String APPEND_PUSH_TYPE = "APPEND";
   private static final String REFRESH_PUSH_TYPE = "REFRESH";
   private static final String EPOCH_TIME_FORMAT = "EPOCH";
   private static final String SIMPLE_DATE_TIME_FORMAT = "SIMPLE_DATE_FORMAT";
   private static final String LONG_SIMPLE_DATE_FORMAT = "yyyyMMdd";
   private static final String STRING_SIMPLE_DATE_FORMAT = "yyyy-MM-dd";
+  private static final String STRING_SLASH_DATE_FORMAT = "yyyy/MM/dd";
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/test/java/org/apache/pinot/segment/spi/creator/name/NormalizedDateSegmentNameGeneratorTest.java:40]: `MALFORMED_SIMPLE_DATE_FORMAT`?

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-segment-spi/src/test/java/org/apache/pinot/segment/spi/creator/name/NormalizedDateSegmentNameGeneratorTest.java (hunk 3 above)
Comment (walterddr) [reply pinot-segment-spi/src/test/java/org/apache/pinot/segment/spi/creator/name/NormalizedDateSegmentNameGeneratorTest.java:40]: not necessarily, SLASH_DATE_FORMAT is acceptable for NormalizedDateSegmentNameGenerator

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/SimpleSegmentNameGenerator.java (hunk 2 above)
Comment (walterddr) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/SimpleSegmentNameGenerator.java:54]: dont think there's a rule for checking this. will update checkstyle.xml in separate PR

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/FixedSegmentNameGenerator.java [hunk 4]
```diff
@@ -28,6 +29,8 @@
   private final String _segmentName;
 
   public FixedSegmentNameGenerator(String segmentName) {
+    Preconditions.checkArgument(
+        segmentName != null && isValidSegmentName(segmentName));
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/FixedSegmentNameGenerator.java:33]: (nit) can be on the same line?

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/NormalizedDateSegmentNameGenerator.java [hunk 5]
```diff
@@ -53,6 +53,8 @@ public NormalizedDateSegmentNameGenerator(String tableName, @Nullable String seg
       boolean excludeSequenceId, @Nullable String pushType, @Nullable String pushFrequency,
       @Nullable DateTimeFormatSpec dateTimeFormatSpec) {
     _segmentNamePrefix = segmentNamePrefix != null ? segmentNamePrefix.trim() : tableName;
+    Preconditions.checkArgument(
+        _segmentNamePrefix == null || isValidSegmentName(_segmentNamePrefix));
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/NormalizedDateSegmentNameGenerator.java:57]:
```suggestion
        _segmentNamePrefix != null && isValidSegmentName(_segmentNamePrefix));
```

Comment (walterddr) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/NormalizedDateSegmentNameGenerator.java]: good catch. forgot to change after merging the 2 rules.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/FixedSegmentNameGenerator.java [hunk 6]
```diff
@@ -28,6 +29,8 @@
   private final String _segmentName;
 
   public FixedSegmentNameGenerator(String segmentName) {
+    Preconditions.checkArgument(
```
Comment (walterddr) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/FixedSegmentNameGenerator.java:32]:
```suggestion
    Preconditions.checkArgument(segmentName != null && isValidSegmentName(segmentName));
```

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/FixedSegmentNameGenerator.java (hunk 4 above)
Comment (walterddr) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/name/FixedSegmentNameGenerator.java:33]:
```suggestion
```

**Pull request merged**